File: Backend/TrackrAPI/TrackrAPI/Services/Inventario/AlmacenService.cs
using TrackrAPI.Dtos.Inventario;
using TrackrAPI.Models;
using TrackrAPI.Repositorys.Inventario;
using System.Collections.Generic;
using System.Transactions;

namespace TrackrAPI.Services.Inventario
{
    public class AlmacenService
    {
        private IAlmacenRepository almacenRepository;
        private IUbicacionRepository ubicacionRepository;
        private IUsuarioAlmacenRepository usuarioAlmacenRepository;
        private AlmacenValidatorService almacenValidatorService;

        public AlmacenService(IAlmacenRepository almacenRepository,
            IUbicacionRepository ubicacionRepository,
            IUsuarioAlmacenRepository usuarioAlmacenRepository,
            AlmacenValidatorService almacenValidatorService)
        {
            this.almacenRepository = almacenRepository;
            this.ubicacionRepository = ubicacionRepository;
            this.usuarioAlmacenRepository = usuarioAlmacenRepository;
            this.almacenValidatorService = almacenValidatorService;
        }

        public AlmacenDto ConsultarDto(int idAlmacen)
        {
            return almacenRepository.ConsultarDto(idAlmacen);
        }

        public IEnumerable<AlmacenGridDto> ConsultarGeneral(int idUsuario)
        {
            return almacenRepository.ConsultarGeneral(idUsuario);
        }

        public IEnumerable<AlmacenDto> ConsultarPorCompania(int idCompania, int idUsuario)
        {
            return almacenRepository.ConsultarPorCompania(idCompania, idUsuario);
        }

        public IEnumerable<AlmacenGridDto> ConsultarPorEstado(int idEstado)
        {
            return almacenRepository.ConsultarPorEstado(idEstado);
        }

        public IEnumerable<AlmacenGridDto> ConsultarPorEstatus(int idEstatusAlmacen)
        {
            return almacenRepository.ConsultarPorEstatus(idEstatusAlmacen);
        }

        public IEnumerable<AlmacenGridDto> ConsultarPorUsuario(int idUsuarioResponsable)
        {
            return almacenRepository.ConsultarPorUsuario(idUsuarioResponsable);
        }

        public IEnumerable<AlmacenDto> ConsultarTodosParaSelector(int idCompania)
        {
            return almacenRepository.ConsultarTodosParaSelector(idCompania);
        }

        public int Agregar(Almacen almacen)
        {
            using (TransactionScope scope = new TransactionScope())
            {
                almacenValidatorService.ValidarAgregar(almacen);

                var idAlmacen = almacenRepository.Agregar(almacen).IdAlmacen;

                var ubicacionDefault = new Ubicacion();
                ubicacionDefault.Nombre = "General";
                ubicacionDefault.IdAlmacen = almacen.IdAlmacen;

                ubicacionRepository.Agregar(ubicacionDefault);

                scope.Complete();

                return idAlmacen;
            }
        }

        public void Editar(Almacen almacen)
        {
            almacenValidatorService.ValidarEditar(almacen);

            almacenRepository.Editar(almacen);
        }

        public void Eliminar(int idAlmacen)
        {
            using (TransactionScope scope = new TransactionScope())
            {
                Almacen almacen = almacenRepository.Consultar(idAlmacen);

                almacenValidatorService.ValidarEliminar(idAlmacen);

                var usuariosAlmacen = usuarioAlmacenRepository.ConsultarPorAlmacen(idAlmacen);

                usuarioAlmacenRepository.Eliminar(usuariosAlmacen);

                almacenRepository.Eliminar(almacen);

                scope.Complete();
            }
        }
    }
}

File: Frontend/trackr-administrador/src/app/shared/http/inventario/almacen.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Almacen } from '@models/inventario/almacen';

@Injectable()
export class AlmacenService {
  private dataUrl = 'almacen/';

  constructor(public http: HttpClient) {}

  consultar(idAlmacen: number) {
    return this.http.get<Almacen>(this.dataUrl + `consultar/${idAlmacen}`);
  }

  consultarGeneral() {
    return this.http.get<Almacen[]>(this.dataUrl + 'consultarGeneral');
  }
  consultarTodosParaSelector() {
    return this.http.get<Almacen[]>(this.dataUrl + 'consultarTodosParaSelector');
  }

  consultarPorEstado(idEstado: number) {
    return this.http.get<Almacen[]>(this.dataUrl + `consultarPorEstado/${idEstado}`);
  }

  consultarPorRol(idResponsable: number) {
    return this.http.get<Almacen[]>(this.dataUrl + `consultarPorUsuario/${idResponsable}`);
  }

  consultarPorEstatus(idEstatus: number) {
    return this.http.get<Almacen[]>(this.dataUrl + `consultarPorEstatus/${idEstatus}`);
  }

  consultarPorCompania() {
    return this.http.get<Almacen[]>(this.dataUrl + 'consultarPorCompania');
  }

  agregar(almacen: Almacen) {
    return this.http.post<number>(this.dataUrl + 'agregar', almacen);
  }

  editar(almacen: Almacen) {
    return this.http.put<void>(this.dataUrl + 'editar', almacen);
  }

  eliminar(idAlmacen: number) {
    return this.http.delete<void>(this.dataUrl + `eliminar/${idAlmacen}`);
  }
}
